linera-core/src/worker.rs

```diff
@@ -726,14 +763,55 @@ where
             notify_when_messages_are_delivered,
         )
         .await;
-        self.cache_recent_value(Cow::Owned(certificate.value)).await;
+        self.cache_recent_hashed_certificate_value(Cow::Owned(certificate.value))
+            .await;
 
         #[cfg(with_metrics)]
         NUM_BLOCKS.with_label_values(&[]).inc();
 
         Ok((info, actions))
     }
 
+    /// Returns an error if the block requires a blob we don't have, or if unrelated blobs were provided.
+    async fn check_no_missing_blob(
+        &self,
+        block: &Block,
+        hashed_blobs: &[HashedBlob],
+    ) -> Result<(), WorkerError> {
+        let mut required_blob_ids = BTreeSet::new();
+        // Find all certificates containing blobs used when executing this block.
+        for operation in &block.operations {
+            if let Operation::System(SystemOperation::PublishBlob { blob_id }) = operation {
+                required_blob_ids.insert(blob_id);
+            }
+        }
+
+        let mut all_required_blob_ids = required_blob_ids.clone();
+        for hashed_blob in hashed_blobs {
+            let blob_id = hashed_blob.id();
+            ensure!(
+                all_required_blob_ids.remove(&blob_id),
+                WorkerError::UnneededBlob { blob_id }
+            );
+        }
+
+        let blob_ids: HashSet<BlobId> = hashed_blobs
+            .iter()
+            .map(|hashed_blob| hashed_blob.id())
+            .collect();
```

Comment:
We don't need that again, if we just reuse `all_required_blob_ids`: We already removed the provided ones from there.